File: android/app/src/main/java/com/alangpierce/lambdacalculusplayground/definition/ExpressionNumbers.java
package com.alangpierce.lambdacalculusplayground.definition;

import com.alangpierce.lambdacalculusplayground.expression.Expression;
import com.alangpierce.lambdacalculusplayground.expression.Expressions;
import com.alangpierce.lambdacalculusplayground.expression.FuncCall;
import com.alangpierce.lambdacalculusplayground.expression.Lambda;
import com.alangpierce.lambdacalculusplayground.expression.Variable;
import com.alangpierce.lambdacalculusplayground.userexpression.UserExpression;

import javax.annotation.Nullable;

public class ExpressionNumbers {
    /**
     * If this expression is a number, return a string with that number. For example,
     * L s[S z[s(s(z))]] is converted to "2". Note that the variables must be named "s" and "z".
     *
     * If the expression is not a number, returns null.
     */
    public static String tryResolveToNumber(Expression expression) {
        if (!(expression instanceof Lambda)) {
            return null;
        }
        Lambda lambda = (Lambda) expression;
        if (!lambda.varName().equals("s")) {
            return null;
        }
        expression = lambda.body();
        if (!(expression instanceof Lambda)) {
            return null;
        }
        lambda = (Lambda) expression;
        if (!lambda.varName().equals("z")) {
            return null;
        }
        expression = lambda.body();
        int result = 0;
        while (expression instanceof FuncCall) {
            FuncCall funcCall = ((FuncCall) expression);
            if (!funcCall.func().equals(Variable.create("s"))) {
                return null;
            }
            expression = funcCall.arg();
            result++;
        }
        if (!expression.equals(Variable.create("z"))) {
            return null;
        }
        return Integer.toString(result);
    }

    public static boolean isNumber(String defName) {
        try {
            int value = Integer.parseInt(defName);
            // We only allow natural (non-negative) numbers in canonical form. So -3 and 05 are both
            // disallowed.
            return value >= 0 && Integer.toString(value).equals(defName);
        } catch (NumberFormatException e) {
            return false;
        }
    }

    /**
     * If the given string is a number, return an expression for it. Otherwise, return null;
     */
    public static @Nullable Expression tryExpressionForNumber(String number) {
        if (!isNumber(number)) {
            return null;
        }
        int value = Integer.parseInt(number);

        Expression result = Variable.create("z");
        for (int i = 0; i < value; i++) {
            result = FuncCall.create(Variable.create("s"), result);
        }
        result = Lambda.create("s", Lambda.create("z", result));
        return result;
    }

    public static @Nullable UserExpression tryUserExpressionForNumber(String number)
            throws ExpressionTooBigException {
        // Refuse to show the user the expanded form for gigantic numbers, since it'll cause
        // performance issues.
        if (isNumber(number) && Integer.parseInt(number) > 30) {
            throw new ExpressionTooBigException();
        }

        @Nullable Expression expression = tryExpressionForNumber(number);
        if (expression == null) {
            return null;
        }
        return Expressions.toUserExpression(expression);
    }
}

File: ExpressionNumbers.js
/**
 * @flow
 */

import type {
    Expression,
    UserExpression,
} from './types';
import {
    FuncCall,
    Lambda,
    UserFuncCall,
    UserLambda,
    UserVariable,
    Variable,
} from './types';

export const tryExpressionForNumber = (defName: string): ?Expression => {
    if (!isNumber(defName)) {
        return null;
    }
    const num = Number(defName);
    let body = Variable.make('z');
    for (let i = 0; i < num; i++) {
        body = FuncCall.make(Variable.make('s'), body);
    }
    return Lambda.make('s', Lambda.make('z', body));
};

export const tryUserExpressionForNumber = (defName: string): ?UserExpression => {
    if (!isNumber(defName)) {
        return null;
    }
    const num = Number(defName);
    let body = UserVariable.make('z');
    for (let i = 0; i < num; i++) {
        body = UserFuncCall.make(UserVariable.make('s'), body);
    }
    return UserLambda.make('s', UserLambda.make('z', body));
};

export const isNumber = (defName: string): boolean => {
    const num = Number(defName);
    // The name must be a natural number in canonical form, so -5 and 03 are
    // both disallowed.
    return num >= 0 && ('' + num == defName);
};

/**
 * Return true if the given number is too big.
 */
export const isGiantNumber = (defName: string): boolean => {
    return isNumber(defName) && Number(defName) > 30;
};

export const tryResolveToNumber = (expr: Expression): ?string => {
    if (!(expr instanceof Lambda) || expr.varName !== 's') {
        return null;
    }
    expr = expr.body;
    if (!(expr instanceof Lambda) || expr.varName !== 'z') {
        return null;
    }
    expr = expr.body;
    let result = 0;
    while (expr instanceof FuncCall) {
        const {func, arg} = expr;
        if (!(func instanceof Variable) || func.varName !== 's') {
            return null;
        }
        expr = arg;
        result++;
    }
    if (!(expr instanceof Variable) || expr.varName !== 'z') {
        return null;
    }
    return '' + result;
};